yarl/_url.py

```diff
@@ -17,16 +30,27 @@
 
 sentinel = object()
 
+_SimpleQuery = Union[str, int, float]
+_QueryVariable = Union[_SimpleQuery, "Sequence[_SimpleQuery]"]
+_Query = Union[
+    None, str, "Mapping[str, _QueryVariable]", "Sequence[Tuple[str, _QueryVariable]]"
+]
+
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    Self = Any
```

Comment:
I think it's common to rely on [`typing-extensions`](https://typing-extensions.rtfd.io/en/latest/#typing_extensions.Self) in such cases… It'd need to go into `setup.cfg` with an env marker.

Reply:
There has been an effort to remove typing-extensions https://github.com/aio-libs/aiohttp/pull/8352

Reply:
I think that's because it was no longer used. I don't know why we'd want to ban it for good. Perhaps @Dreamsorcerer could explain..

Reply:
I think it's a good idea to avoid reintroducing it as a dependency, many users prefer fewer dependencies.
Typing in Python has reached a point where it's not really needed for normal use cases, newer versions of Python have small improvements, but nothing that really makes a difference to the end user.

Note this can also be done with a TypeVar, but given that it's basically only the oldest 2 versions of Python that would see the Any, most users will end up type checking with mypy against a more recent release (even if runtime might be on an older release).

Also, are we intending to ditch the .pyi files? If not, then none of this is even exposed to the users...

Reply:
> Also, are we intending to ditch the .pyi files? If not, then none of this is even exposed to the users...

I had originally marked this as `contrib` only because of that, but it probably make more sense to ditch the `pyi` instead of trying to keep them in sync

Reply:
My understanding is that the pyi files are needed for C-extensions. Is Cython capable of producing the proper annotations for MyPy to consume?

Reply:
We need the pyi for the C-extensions, but it looks like everything that in the `__init__.pyi` `__all__` is pure python unless I've overlooked something

Reply:
Agreed.

Reply:
We do need to pull in the typed dict for the cache here as well.. I'll do that after #1086 and remove the `__init__.pyi` so we can rely on the inline types instead